caas/kubernetes/provider/rbac.go

```diff
@@ -0,0 +1,463 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"github.com/juju/errors"
+	core "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/juju/juju/caas"
+)
+
+func (k *kubernetesClient) getRBACLabels(appName string) map[string]string {
+	return map[string]string{
+		labelApplication: appName,
+		labelModel:       k.namespace,
+	}
+}
+
+func (k *kubernetesClient) ensureServiceAccountForApp(
+	appName string, caasSpec *caas.ServiceAccountSpec,
+) (cleanups []func(), err error) {
+	labels := k.getRBACLabels(appName)
+	saSpec := &core.ServiceAccount{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      caasSpec.Name,
+			Namespace: k.namespace,
+			Labels:    labels,
+		},
+		AutomountServiceAccountToken: caasSpec.AutomountServiceAccountToken,
+	}
+	// ensure service account;
+	sa, err := k.updateServiceAccountForApp(appName, saSpec)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return cleanups, errors.Trace(err)
+		}
+		if sa, err = k.createServiceAccount(saSpec); err != nil {
+			return cleanups, errors.Trace(err)
+		}
+		cleanups = append(cleanups, func() { k.deleteServiceAccount(sa.GetName()) })
+	}
+
+	roleSpec := caasSpec.Capabilities.Role
+	var roleRef rbacv1.RoleRef
+	switch roleSpec.Type {
+	case caas.Role:
+		var r *rbacv1.Role
+		if len(roleSpec.Rules) == 0 {
+			// no rules was specified, reference to an existing Role.
+			r, err = k.getRole(roleSpec.Name)
+			if err != nil {
+				return cleanups, errors.Trace(err)
+			}
+		} else {
+			// create or update Role.
+			r, err = k.ensureRole(&rbacv1.Role{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      roleSpec.Name,
+					Namespace: k.namespace,
+					Labels:    labels,
+				},
+				Rules: roleSpec.Rules,
+			})
+			if err != nil {
+				return cleanups, errors.Trace(err)
+			}
+			cleanups = append(cleanups, func() { k.deleteRole(r.GetName()) })
+		}
+		roleRef = rbacv1.RoleRef{
+			Name: r.GetName(),
+			Kind: string(roleSpec.Type),
+		}
+	case caas.ClusterRole:
+		var cr *rbacv1.ClusterRole
+		if len(roleSpec.Rules) == 0 {
+			// no rules was specified, reference to an existing ClusterRole.
+			cr, err = k.getClusterRole(roleSpec.Name)
+			if err != nil {
+				return cleanups, errors.Trace(err)
+			}
+		} else {
+			// create or update ClusterRole.
+			cr, err = k.ensureClusterRole(&rbacv1.ClusterRole{
```

Comment:
This seems like the same thing but since it's cluster-wide it's more likely to be an issue.